torch/csrc/jit/passes/annotate_effects.cpp

```diff
@@ -0,0 +1,227 @@
+#include "torch/csrc/jit/passes/annotate_effects.h"
+
+#include <set>
+#include "torch/csrc/jit/passes/dead_code_elimination.h"
+
+namespace torch {
+namespace jit {
+namespace {
+
+/**
+ * AnnotateEffects
+ *
+ * This pass annotates effectful operations (such as ones that mutate existing
+ * values) to prevent subsequent passes from re-ordering ops in a way that
+ * changes the meaning of the program.
+ *
+ * It does this by threading a "world token" value through nodes that use
+ * mutable values. This models effects explicitly in the IR and forces all
+ * annotated nodes to be linearized during optimization.
+ *
+ * The world token is threaded directly through any nodes that mutate values.
+ * For purely functional operators, their node will be "fenced" by two
+ * `prim::MemoryFence` nodes that take world tokens as their input.
+ */
+class AnnotateEffectsImpl {
+ public:
+  void annotateEffects(Graph* g) {
+    // TODO(suo): We need to change this so that the world token is an input
+    // and output of the graph. That would require changing a bunch of interop
+    // and batching code, so leaving it out for now.
+    //
+    // auto curToken = g->addInput("world")->setType(WorldType::get());
+    // curToken = visitBlock(g->block(), curToken);
+    // g->registerOutput(curToken);
+
+    // Generate the first world token
+    const auto tokenGenerator = g->create(prim::Constant);
+    g->block()->prependNode(tokenGenerator);
+    auto curToken = tokenGenerator->output()->setType(WorldType::get());
+
+    visitBlock(g->block(), curToken);
+  }
+
+ private:
+  Value* visitBlock(Block* block, Value* curToken) {
+    for (auto* node : block->nodes()) {
+      curToken = visitNode(node, curToken);
+    }
+    return curToken;
+  }
+
+  // If a node uses a mutable variable (or mutates a previously constant
+  // variable), annotate it
+  //
+  // Returns the last world token emitted for subsequent annotations to use.
+  Value* visitNode(Node* node, Value* curToken) {
+    if (node->kind() == prim::MemoryFence) {
```

Comment:
Once we insert a memory fence, the next call to `visitNode` visits that memory fence node (since it gets added to the block's node list). This check is to avoid an infinite loop; I'll add a comment to clarify.